docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,518 @@
+# MeshExternalService API
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6330
+
+## Context and Problem Statement
+
+The current implementation of `ExternalService` has significant limitations and warrants a complete redesign.
+
+* Lack of support for * subdomains
+* Lack of support for thick clients (think Kafka or C* where discovery is done as part of the client).
+* Need for pluggability (for example lambda support or equivalent)
+* Working well with gateways (what if your ExternalService relies on SNI/host header to route through a gateway)
+* Applying policies with and without egress
+
+
+## Considered Options
+
+* Creating MeshExternalService + HostnameGenerator
+* Creating MeshExternalService
+* Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+## Decision Outcome
+
+Chosen option: "Creating MeshExternalService".
+
+### Positive Consequences
+
+* Clearer structure.
+* Enhanced functionalities.
+* Better policy matching capabilities.
+* No need for a HostnameGenerator
+
+### Negative Consequences
+
+* Whole new resource with a lot of new code.
+* Temporarily increased complexity of a product until the migration is done.
+* Problem with supporting Universal without transparent proxy, because one MeshExternalService points to many real domains
+
+## Pros and Cons of the Options
+
+### Creating MeshExternalService
+
+We're introducing a new object called `MeshExternalService`. This object will be a namespace resource and will only be creatable within the `kuma-system` namespace. It can be created within a specific zone, but in this case, it will only be accessible within the zone where it was created.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP # Kuma will generate a domain
```

Comment:
What are use cases for having multiple entries? Would it be simpler to have a single one?

Reply:
From an implementation standpoint, it would be simpler to have a single match entry. However, I can see a use case where someone might want to expose multiple domains through passthrough without creating many external services.